index.js

```diff
@@ -3,6 +3,35 @@ var RSVP = require('rsvp');
 var exit;
 var handlers = [];
 var lastTime;
+var isExiting = false;
+
+process.on('beforeExit', function (code) {
+  if (handlers.length === 0) { return; }
+
+  var own = lastTime = module.exports._flush(lastTime, code)
+    .finally(function () {
+      // if an onExit handler has called process.exit, do not disturb
+      // `lastTime`.
+      //
+      // Otherwise, clear `lastTime` so that we know to synchronously call the
+      // real `process.exit` with the given exit code, when our captured
+      // `process.exit` is called during a `process.on('exit')` handler
+      //
+      // This is impossible to reason about, don't feel bad.  Just look at
```

Comment:
Maybe I should update it to say it's only impossible for mere mortals to reason about ^_^